### PR TITLE
feat(mcp): add OAuth redirect URI configuration for MCP servers

### LINKED ISSUE #7377: [FEATURE]: Allow custom OAuth redirect URI configuration for MCP servers
## Summary

This feature request is focused on adding support for **custom OAuth redirect URI configuration** for MCP servers.

## Problem

OpenCode currently hardcodes the OAuth callback URL as `http://127.0.0.1:19876/mcp/oauth/callback`. Many OAuth servers (Azure AD, Okta, Keycloak, GitHub OAuth apps, etc.) require exact matching of pre-registered redirect URIs, which prevents users from:
- Using OpenCode with enterprise OAuth providers that use different domains
- Configuring OAuth for deployment scenarios with custom hostnames
- Using Claude Code with different callback port configurations

## Design Philosophy

**OpenCode already allows overriding the `clientId`** to impersonate different OAuth clients. Since OAuth servers validate **both** the client ID and the redirect URI as part of the security model, it follows logically that users should also be able to override the **redirect URI**.

Without this capability, there's an asymmetry where users can point to any OAuth server and any client, but are still constrained by OpenCode's hardcoded callback URL. This defeats the purpose of flexible client configuration.

**Therefore, supporting custom `redirectUri` is essential for feature completeness** when users have the ability to override `clientId`.

## Solution

Add a new optional `redirectUri` configuration parameter to MCP OAuth settings, allowing users to specify custom callback URLs:

```jsonc
{
  "mcp": {
    "my-server": {
      "type": "remote",
      "url": "https://api.example.com",
      "oauth": {
        "clientId": "my-client-id",           // Already configurable
        "redirectUri": "https://custom.example.com/callback"  // NEW: Also configurable
      }
    }
  }
}
```

## Implementation Details

- Add optional `redirectUri` field to MCP OAuth config schema
- Update OAuth provider to use configured URI when available
- Dynamically configure callback server port/path based on parsed redirect URI
- Fall back to defaults (`http://127.0.0.1:19876/mcp/oauth/callback`) when not specified
- Full test coverage for custom URI parsing and edge cases

## Backward Compatibility

✅ Fully backward compatible - when `redirectUri` is not specified, uses the default OpenCode callback URL.

## Related Issues

This feature was originally part of #5766 which also addressed MCP SDK version updates for authorization endpoint discovery.

Splitting these concerns:
- **#5766**: MCP SDK upgrade (authorization endpoint fix)
- **This issue**: Custom redirect URI support

### PR DESCRIPTION
**Pull request by christso**
## Summary

Re-implements the `redirectUri` option for MCP OAuth configuration, allowing users to specify a custom callback URL for OAuth flows.

- Adds `redirectUri` config option to `McpOAuth` schema
- Updates `McpOAuthProvider` to use custom redirect URI when provided
- Updates `McpOAuthCallback.ensureRunning()` to support custom port/path
- Adds `parseRedirectUri()` utility function

**Key difference from original PR #7379:** The OAuth callback server is now started lazily in `startAuth()` only when authentication is actually needed, rather than preemptively in `create()` for all remote servers. This fixes the regression where non-OAuth MCP servers would fail to connect.

Closes #7377

## Test plan

- [x] Existing MCP tests pass (6 tests)
- [x] New `parseRedirectUri` tests pass (3 tests)
- [x] New `ensureRunning` with custom redirectUri test passes (1 test)
- [x] Manual test: Context7 remote MCP server connects without OAuth callback server
- [x] Manual test: Local filesystem MCP server connects normally

### Manual test results

```
┌  MCP Servers
│
●  ✓ context7 connected
│      https://mcp.context7.com/mcp
│
●  ✓ filesystem connected
│      npx -y @modelcontextprotocol/server-filesystem /tmp
│
└  2 server(s)
```

### CODE REVIEW

**Comment by egze** [comment]
Thanks for the PR. I am directly affected that I can't set a custom redirectUri.

**Comment by christso** [comment]
Here's the configuration that Claude Code supports when I run `claude mcp add -h`. OpenCode should also support defining the callback port or URI.

```
Options:
  --callback-port <port>       Fixed port for OAuth callback (for servers requiring pre-registered redirect URIs)
  --client-id <clientId>       OAuth client ID for HTTP/SSE servers
  --client-secret              Prompt for OAuth client secret (or set MCP_CLIENT_SECRET env var)
  -e, --env <env...>           Set environment variables (e.g. -e KEY=value)
  -H, --header <header...>     Set WebSocket headers (e.g. -H "X-Api-Key: abc123" -H "X-Custom: value")
  -h, --help                   Display help for command
  -s, --scope <scope>          Configuration scope (local, user, or project) (default: "local")
  -t, --transport <transport>  Transport type (stdio, sse, http). Defaults to stdio if not specified.
```

It's been a while since I submitted the PR because I was happy with the workaround of running a script to copy the MCP token from CC to OpenCode. But it seems a lot of my colleagues are not happy with the workaround, so I'm revisiting this.

### Risk analysis

This is a v2 of #7379, which was reverted due to a regression where the OAuth callback server was started eagerly in `create()` for all remote servers, causing non-OAuth MCP servers (e.g. Context7) to fail.

**The regression is fixed in this version.** The key change is lazy initialization -- `ensureRunning()` is now only called inside `startAuth()`, which is only triggered when OAuth authentication is actually needed. Non-OAuth servers never call `startAuth()`, so they are unaffected.

**Theoretical race condition in the stop/restart logic is not a practical concern.** The code stops and restarts the callback server when `ensureRunning()` is called with a different port/path, which clears all pending OAuth callbacks. However, `authenticate()` is user-initiated and sequential -- a user would have to be mid-OAuth-flow for one server and simultaneously trigger OAuth for a second server with a *different* `redirectUri`. The single-threaded event loop and the interactive nature of OAuth (browser redirect, user action) make this effectively impossible in normal usage.

Ready to merge. @rekram1-node

File: packages/opencode/src/mcp/oauth-callback.ts [hunk 1]
```diff
@@ -56,21 +60,33 @@ export namespace McpOAuthCallback {
...
-  export async function ensureRunning(): Promise<void> {
+  export async function ensureRunning(redirectUri?: string): Promise<void> {
+    // Parse the redirect URI to get port and path (uses defaults if not provided)
+    const { port, path } = parseRedirectUri(redirectUri)
+
+    // If server is running on a different port/path, stop it first
+    if (server && (currentPort !== port || currentPath !== path)) {
+      log.info("stopping oauth callback server to reconfigure", { oldPort: currentPort, newPort: port })
+      await stop()
+    }
```
Comment (christso) [inline packages/opencode/src/mcp/oauth-callback.ts]: Lazy initialization fix for the regression in #7379 -- this stop/restart path is only reachable during an active OAuth flow (via `startAuth()`), so non-OAuth servers are never affected. The theoretical race with concurrent different-redirectUri auths is not a practical concern since `authenticate()` is user-initiated and sequential.

**Comment by dbpolito** [comment]
would love to see this one merged to use github mcp with oauth

**Comment by chandlerok** [comment]
would love to see this merged

**Comment by benceferdinandy-signifyd** [comment]
@christso seems this ran ontu a merge conflict. would be nice to see this merged ..

**Comment by egze** [comment]
@christso I made a new PR based on yours https://github.com/anomalyco/opencode/pull/21385

If you still want to make the changes in your branch - just tell me and I'll close mine. I just need this feature so badly 😄

**Comment by mzyil** [comment]
any updates on this?

**Comment by christso** [comment]
Superseded by #21385 — closing this PR.

Thanks @egze for picking this up and re-implementing it for the current codebase! The Node http.createServer adaptation was the right call given the upstream changes.
